Use ~~pycrypto~~ os.urandom() instead of pyNacl for random generation.

rspier opened the pull request:
pycrypto allows this library to be used on App Engine (which doesn't support
ffi which is required for pyNacl.)

Bennyoak commented:
Thanks a ton for your PR! Sorry it's taken so long to respond. New job has taken all my time.

I'm hoping to get back into some much needed updates (among them, making pynacl optional, or at lease deprecated). This will be a tremendous help. Thank you!

rspier commented:
No rush!
Happy to be helpful.

alanfairless commented:
Could you just use `os.urandom()` from the standard library instead, and avoid the dependency altogether?

rspier commented:
After a little bit of digging, I'm pretty convinced the answer is yes.  No one seems to have any concerns about it's randomness, and the biggest potential issue is a read that blocks -- which is extremely unlikely to happen in a real world scenario.

It's @Bennyoak's call.  I'll happily update the pull request to os.urandom()

Bennyoak commented:
My initial reason for not using `os.urandom()` was over concerns that it wouldn't be secure enough. I admit I may have been a bit over zealous at the time :)

That said, I've always toy'd with the idea of having options for a greater range of support. Somebody with a run-of-the-mill site may be just fine using urandom, whereas a larger site with more at stake may insist on a _stronger_ cryptographic engine, like pycrypto or PyNacl.

In that spirit, I propose we use both! @rspier @alanfairless - what are your thoughts on making `os.urandom()` a default, with `pycrypto` available as a settings switch?

rspier commented:
I'm not a crypto expert, but I think that the flexibility might be unnecessary.  I couldn't find any evidence that pycrypto or PyNacl are better CPRNGs than /dev/urandom (at least on Linux). I even found some evidence that pycrypto's implementation might not be the greatest (or at least isn't as well studied as others.)  On principle, I don't think a config setting is necessarily useful here until someone has a good reason for needing it -- it just adds complexity.

Were we to switch, I'd probably do it through [random.SystemRandom](https://docs.python.org/2/library/random.html#random.SystemRandom).  It's a simple wrapper around os.urandom.

    >>> r = random.SystemRandom()
    >>> r.getrandbits(128)
    298086178031829153016673505629597858480L

Having said "at least on Linux", maybe an optional setting isn't a bad idea if it can be done in a simple manner -- I'd suggest implementing it as "pass a class that implements getrandbits()".

rspier commented:
p.s. links of interest:

  * https://www.2uo.de/myths-about-urandom/
  * https://www.reddit.com/r/crypto/comments/3ghgqt/how_secure_is_the_csprng_inside_devurandom/

(and plenty more)
  
(Yes, these are mostly random people with unknown credentials... but the overwhelming consensus is that urandom is fine.)

Bennyoak commented:
@rspier - Yeah I'm looking through it as well. https://www.2uo.de/myths-about-urandom/ was a good link. Also keeping in mind that we're dealing with nonces and not a stored key, per se. It's beginning to look like `os.urandom()` really is the way to go.

At this point I'm wondering if it's worth catching (and finding a fallback for) `NotImplementedError`. How often is it likely that `dev/urandom` would be unavailable?

Another option mentioned by @kravietz was the [secrets](https://docs.python.org/3/library/secrets.html) module... but I'm not sure if there's a way to backport it to 2.7, or if it's really any better/stronger in this context.

rspier commented:
I don't think it's worth dealing with the NotImplemented error.  It's very unlikely to happen (and all sorts of other things would likely be broken too.)  Looking at the source, it only happens on unix-like systems where /dev/urandom doesn't exist or the user has no permission to read from it.

=secrets= is just a thin wrapper around random.SysRandom: https://github.com/python/cpython/blob/3.6/Lib/secrets.py.  it doesn't seem worth backporting, it's mostly about formatting.

Bennyoak commented:
Cool. `urandom` it is, then!
I'm currently adding Django 1.11 and Python 3.6 to the testing environment. If you want to flip this to use `urandom` feel free. I'll scoop it in with the test updates and make changes accordingly.

rspier commented:
Don't look at that commit, forgot to remove some other pycrypto pieces.

rspier commented:
Sorry for the multiple pushes, I liked the concept of random.SystemRandom, but the reality of the implementation requirements turned me off. 

Internally, random.SystemRandom reads bytes, converts them to hex, converts them back to a long.  I then take that long, converted it back to hex, added padding, converted it to a sequence of bytes, and base 64 encoded it.  Way too many conversions back and forth for little benefit.

Bennyoak commented on README.rst:
```diff
@@ -23,8 +23,6 @@ Disclosure
    has confirmed this doesn’t work with their version.
--  From their logs:
-   ``RuntimeError: PyNaCl is not compatible with PyPy < 2.6. Please upgrade PyPy to use this library.``
 
```
Let's add a notice that DCN no longer supports PyNacl and that newer versions of will not be backwards compatible.

I think I'm going to make this a 1.0.0 push - taking her out of beta :-D

rspier replied:
Won't that confuse users?  DCN never "supported" PyNacl, it was a dependency, and now it's not a dependency.   DCN itself is still backwards compatible because the interface hasn't changed.  If there was a ChangeLog file, the PyNacl removal could go in there.

If anything, maybe consider adding something about attempting to only use "pure Python" dependencies for best compatibility with PyPy and AppEngine?

Bennyoak replied:
Yeah I suppose "support" is the wrong word. I meant just something saying that DCN doesn't use PyNacl anymore. So that anybody updating from the current version can remove it if they don't need it for anything else.

rspier replied:
See 6f0468a below.

Bennyoak replied:
Yeah that's great. Ima rerun the test before merge. Thanks!

Bennyoak commented on csp_nonce/utils.py:
```diff
@@ -1,13 +1,14 @@
-    return "{}".format(base64.b64encode(nonce))
+    """ Return a unique base64 encoded nonce hash."""
+    bits = os.urandom(NONCE_BYTES)
+    return base64.b64encode(bits)
```
Ah - I think we're gonna need this string formatted.
`return "{}".format(base64.b64encode(bits))` should do it.

rspier replied:
Why?

    >>> import base64
    >>> "{}".format(base64.b64encode("12345"))
    'MTIzNDU='
    >>> base64.b64encode("12345")
    'MTIzNDU='

rspier replied:
If the answer is Python 3, we might want to keep digging:

Python 2:
    >>> "{}".format(base64.b64encode(os.urandom(3)))
    '7Fr+'

Python 3:
    >>> "{}".format(base64.b64encode(os.urandom(3)))
    "b'9Uq9'"

Notice it's outputing the repr value.  I think what we're looking for is actually .decode()

Python 2:
    >>> base64.b64encode(os.urandom(3)).decode()
    u'rVvX'

Python 3:
    >>> base64.b64encode(os.urandom(3)).decode()
    'gQT6'

Bennyoak replied:
Hah - beat me to it.
Lets go try that and see if the tests pass.

rspier replied:
It's looking better already.

Bennyoak commented:
Bam! Great job! Thanks again for your contribution.

rspier commented:
My pleasure.  Thanks!
